filebeat/channel/runner_mock_test.go

```diff
@@ -0,0 +1,118 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package channel
+
+import (
+	"testing"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/cfgfile"
+
+	conf "github.com/elastic/elastic-agent-libs/config"
+
+	"github.com/stretchr/testify/require"
+)
+
+type runnerFactoryMock struct {
+	cfgfile.RunnerFactory
+	clientCount int
+	cfgs        []beat.ClientConfig
+}
+
+func (r *runnerFactoryMock) Create(p beat.PipelineConnector, config *conf.C) (cfgfile.Runner, error) {
+	// When using the connector multiple times to create a client
+	// it's using the same editor function for creating a new client
+	// with a modified configuration that includes predefined processing.
+	// This is why we must make sure nothing is re-used from one client to another.
+	for i := 0; i < r.clientCount; i++ {
+		client, err := p.ConnectWith(beat.ClientConfig{})
+		if err != nil {
+			return nil, err
+		}
+
+		// storing the config that the client was created with
+		// it's needed for the `Assert` later
+		r.cfgs = append(r.cfgs, client.(*clientMock).cfg)
+	}
+	return &struct {
+		cfgfile.Runner
+	}{}, nil
+}
+
+// Assert runs various checks for the clients created by the wrapped pipeline connector
+// We check that the processing configuration does not reference the same addresses as before,
+// re-using some parts of the processing configuration will result in various issues, such as:
+// * closing processors multiple times
+// * using closed processors
+// * modifiying an object shared by multiple pipeline clients
+func (r runnerFactoryMock) Assert(t *testing.T) {
+	t.Helper()
+
+	// we need to make sure `Assert` is called after `Create`
+	require.Len(t, r.cfgs, r.clientCount)
+
+	t.Run("new processing configuration each time", func(t *testing.T) {
+		for i, c1 := range r.cfgs {
+			for j, c2 := range r.cfgs {
+				if i == j {
+					continue
+				}
+
+				require.NotSamef(t, c1.Processing, c2.Processing, "processing configuration cannot be re-used")
+				require.NotSamef(t, c1.Processing.Meta, c2.Processing.Meta, "`Processing.Meta` cannot be re-used")
+				require.NotSamef(t, c1.Processing.Fields, c2.Processing.Fields, "`Processing.Fields` cannot be re-used")
+				require.NotSamef(t, c1.Processing.Processor, c2.Processing.Processor, "`Processing.Processor` cannot be re-used")
+			}
+		}
+	})
+
+	t.Run("new processors each time", func(t *testing.T) {
+		var processors []beat.Processor
+		for _, c := range r.cfgs {
+			processors = append(processors, c.Processing.Processor.All()...)
+		}
+
+		require.NotEmptyf(t, processors, "for this test the list of processors cannot be empty")
+
+		for i, p1 := range processors {
+			for j, p2 := range processors {
+				if i == j {
+					continue
+				}
+
+				require.NotSamef(t, p1, p2, "processors must not be re-used")
+			}
+		}
+	})
+}
+
+type clientMock struct {
+	beat.Client
```

Comment:
This case of interface inclusion seems more ok to me, since its only job is to propagate the config data and we can reasonably expect that nothing exercised by these tests will call into it -- because of this it's less brittle to implementation changes. But please add a comment explaining that :-)

Reply:
I don't think we need a comment for each case of embedding an interface. As I replied above, this pattern has a maintainability advantage and it's very concise.

I'd let everything panic, if it panics you know it's not supposed to be used in this test, which is good enough indication for me. It would not panic on the CI unless someone changes a test locally and does not even run it. I hope we never do that.

Reply:
Comments are free. Please don't avoid adding a requested comment just because you think everybody already understands something, especially with a pattern like this that can mask other errors. As I said, I lost multiple hours today to this pattern and there is no possible harm in a one-line comment.

Reply:
Not needed anymore, implemented the interface in https://github.com/elastic/beats/pull/34870/commits/16c10e06c0498694bfd9bed007f0f154fa9c5ba8